### ISSUE TITLE
Quoted annotations should be parsed as if parenthesized

### ISSUE THREAD

**Issue by charliermarsh** [astral-sh/ruff#9467]
It was recently agreed that quoted annotations should be parsed as if they were implicitly wrapped in parentheses. So, e.g., they can contain otherwise-invalid indentation, newlines within binary operators, etc.

See: https://github.com/microsoft/pyright/issues/6940#issuecomment-1886272420

**Comment by Glyphack** [comment]
Should this be implemented in the parser layer or [here](https://github.com/astral-sh/ruff/blob/master/crates/ruff_linter/src/checkers/ast/mod.rs#L1860) where we emit the warniing?

**Comment by dhruvmanila** [comment]
I think it should be implemented in the `parse_type_annotation` function which is used by the AST checker to report an error if found:

https://github.com/astral-sh/ruff/blob/a0263ab472f824b0673a46f6923abe012ba3b374/crates/ruff_python_parser/src/typing.rs#L22-L44

I would also suggest to look at other references of `parse_type_annotation` function to make sure they aren't affected in any way with this change. They shouldn't be but just to make sure :)

**Comment by carljm** [comment]
At this point, we'll want the fix here to apply to both ruff and red-knot.

**Comment by dhruvmanila** [comment]
> At this point, we'll want the fix here to apply to both ruff and red-knot.

Agreed. Both ruff and red knot utilizes the `parse_expression_range` function from the parser:

https://github.com/astral-sh/ruff/blob/21aa12a073091e31fd9ca5fc47a869a257e40df2/crates/ruff_python_parser/src/lib.rs#L157-L167

So, as Micha suggested, we could add a `parse_parenthesized_expression_range` function. I'd suggest adding a new `Mode::ParenthesizedExpression` and initiating the `Lexer` with a `nesting` value of 1 which I think should solve this. It might require some testing.
